dimod/generators/wireless.py

```diff
@@ -0,0 +1,985 @@
+            ...     F_distribution = ('binary', 'real'))
+
+    """
+
+    if not hasattr(lattice, 'edges') or not hasattr(lattice, 'nodes'): # not nx.Graph:
+        raise ValueError('Lattice must be a :class:`networkx.Graph`')
+
+    attenuation_matrix, _, _ = _lattice_to_attenuation_matrix(lattice)
+    
+    num_receivers, num_transmitters = attenuation_matrix.shape
+
+    bqm = mimo(
+        modulation=modulation,
+        y=y,
+        F=F,
+        transmitted_symbols=transmitted_symbols,
+        channel_noise=channel_noise,
+        num_transmitters=num_transmitters,
+        num_receivers=num_receivers,
+        SNRb=SNRb,
+        seed=seed,
+        F_distribution=F_distribution,
+        attenuation_matrix=attenuation_matrix)
+
+    return bqm
+
+# Linear-filter functions. These are not used for spin-encoding MIMO problems
+# and are maintained here for user convenience 
+ 
+def linear_filter(F, method='zero_forcing', SNRoverNt=float('Inf'), PoverNt=1):
+    """Construct a linear filter for estimating transmitted signals.
+
+    Following the conventions of MacKay\ [#Mackay]_, a filter is constructed 
+    for independent and identically distributed Gaussian noise at power spectral 
+    density :math:`N_0`:
+
+    :math:`N_0 I[N_r] = E[n n^{\dagger}]`
+
+    For independent and identically distributed (i.i.d), zero mean, transmitted symbols,
+
+    :math:`P_c I[N_t] = E[v v^{\dagger}]`
+
+    where :math:`P_{c}`, the constellation's mean power, is equal to 
+    :math:`(1, 2, 10, 42)N_t` for BPSK, QPSK, 16QAM, 64QAM respectively.
+
+    For an i.i.d channel,
+
+    :math:`N_r N_t = E_F[T_r[F F^{\dagger}]] \qquad \Rightarrow \qquad E[||F_{\mu, i}||^2] = 1`
+
+    Symbols are assumed to be normalized:
+
+    :math:`\\frac{SNR}{N_t} = \\frac{P_c}{N_0}`
+
+    :math:`SNR_b = \\frac{SNR}{N_t B_c}`
+
+    where :math:`B_c` is bit per symbol, equal to  :math:`(1, 2, 4, 8)`
+    for BPSK, QPSK, 16QAM, 64QAM respectively.
+
+    Typical use case: set :math:`\\frac{SNR}{N_t} = SNR_b`.
+
+    .. [#Mackay] Matthew R. McKay, Iain B. Collings, Antonia M. Tulino.
+        "Achievable sum rate of MIMO MMSE receivers: A general analytic framework"
+        IEEE Transactions on Information Theory, February 2010
+        arXiv:0903.0666 [cs.IT]
+
+    Reference:
+
+        https://www.youtube.com/watch?v=U3qjVgX2poM
+    """
+    if method not in ['zero_forcing', 'matched_filter', 'MMSE']:
+        raise ValueError('Unsupported filtering method')
+
+    if method == 'zero_forcing':
+        # Moore-Penrose pseudo inverse
+        return np.linalg.pinv(F)
+
+    Nr, Nt = F.shape
+
+    if method == 'matched_filter':  # F = root(Nt/P) Fcompconj
+        return F.conj().T / np.sqrt(PoverNt)
+
+    # method == 'MMSE':
+    return np.matmul(
+        F.conj().T,
+        np.linalg.pinv(np.matmul(F, F.conj().T) + np.identity(Nr)/SNRoverNt)
+            ) / np.sqrt(PoverNt)
+
+def filter_marginal_estimator(x: np.array, modulation: str):
```

Comment:
Does this one need a docstring?

Reply:
"""Takes the continuous filter output and maps each symbol estimate to the nearest constellation value (valid symbol)"""

Reply:
Updated, thanks!